File: floodFill.py
from typing import List

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        queue = [(sr, sc)]
        # image[sr][sc] = newColor // we should color (sr, sc) at last
        visited = [(sr, sc)]
        row_shift = [-1, 1, 0, 0]
        col_shift = [0, 0, -1, 1]
        while len(queue) > 0:
            current_row, current_col = queue.pop(0)
            for i, j in zip(row_shift, col_shift):
                neighbor_row = current_row + i
                neighbor_col = current_col + j
                if neighbor_row < 0 or neighbor_row >= len(image) or neighbor_col < 0 or neighbor_col >= len(image[neighbor_row]):
                    continue
                if (not ((neighbor_row, neighbor_col) in visited)) and image[neighbor_row][neighbor_col] == image[sr][sc]:
                    image[neighbor_row][neighbor_col] = newColor
                    visited.append((neighbor_row, neighbor_col))
                    queue.append((neighbor_row, neighbor_col))
        image[sr][sc] = newColor
        return image

if __name__ == "__main__":
    solution = Solution()
    image = [[1, 1, 1], [1, 1, 0], [1, 0, 1]]
    new_image = solution.floodFill(image, 1, 1, 2)
    print(new_image)

File: floodFill.cpp
//
// Created by Golfer on 2020/7/29.
//

#include <vector>
#include <cstdio>
#include <queue>
#include <set>
using namespace std;

class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        queue<pair<int, int>> q;
        pair<int, int> pair_sr_sc = make_pair(sr, sc);
        q.push(pair_sr_sc);
        set<pair<int, int>> visited;
        visited.insert(pair_sr_sc);
        vector<int> row_shift = {-1, 1, 0, 0};
        vector<int> col_shift = {0, 0, -1, 1};
        while (! q.empty()) {
            pair<int, int> current = q.front();
            int current_row = current.first;
            int current_col = current.second;
            q.pop();
            for (int i = 0; i < row_shift.size(); i++) {
                int neighbor_row = current_row + row_shift[i];
                int neighbor_col = current_col + col_shift[i];
                if (neighbor_row < 0 || neighbor_row >= image.size() || neighbor_col < 0 || neighbor_col >= image[neighbor_row].size()) {
                    continue;
                }
                pair<int, int> pair_neighbor = make_pair(neighbor_row, neighbor_col);
                if (image[neighbor_row][neighbor_col] == image[sr][sc] && visited.find(pair_neighbor) == visited.end()) {
                    image[neighbor_row][neighbor_col] = newColor;
                    visited.insert(pair_neighbor);
                    q.push(pair_neighbor);
                }
            }
        }
        image[sr][sc] = newColor;
        return image;
    }
};

